### PR TITLE
[SPARK-7326] [STREAMING] Performing window() on a WindowedDStream doesn't work all the time

### PR DESCRIPTION
**Pull request by wesleymiao**
@tdas 

https://issues.apache.org/jira/browse/SPARK-7326

The problem most likely resides in DStream.slice() implementation, as shown below.

  def slice(fromTime: Time, toTime: Time): Seq[RDD[T]] = {
    if (!isInitialized) {
      throw new SparkException(this + " has not been initialized")
    }
    if (!(fromTime - zeroTime).isMultipleOf(slideDuration)) {
      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
        + slideDuration + ")")
    }
    if (!(toTime - zeroTime).isMultipleOf(slideDuration)) {
      logWarning("toTime (" + fromTime + ") is not a multiple of slideDuration ("
        + slideDuration + ")")
    }
    val alignedToTime = toTime.floor(slideDuration, zeroTime)
    val alignedFromTime = fromTime.floor(slideDuration, zeroTime)

```
logInfo("Slicing from " + fromTime + " to " + toTime +
  " (aligned to " + alignedFromTime + " and " + alignedToTime + ")")

alignedFromTime.to(alignedToTime, slideDuration).flatMap(time => {
  if (time >= zeroTime) getOrCompute(time) else None
})
```

  }

Here after performing floor() on both fromTime and toTime, the result (alignedFromTime - zeroTime) and (alignedToTime - zeroTime) may no longer be multiple of the slidingDuration, thus making isTimeValid() check failed for all the remaining computation.

The fix is to add a new floor() function in Time.scala to respect the zeroTime while performing the floor :

  def floor(that: Duration, zeroTime: Time): Time = {
    val t = that.milliseconds
    new Time(((this.millis - zeroTime.milliseconds) / t) \* t + zeroTime.milliseconds)
  }

And then change the DStream.slice to call this new floor function by passing in its zeroTime.

```
val alignedToTime = toTime.floor(slideDuration, zeroTime)
val alignedFromTime = fromTime.floor(slideDuration, zeroTime)
```

This way the alignedToTime and alignedFromTime are _really_ aligned in respect to zeroTime whose value is not really a 0.

### CODE REVIEW

**Comment by srowen** [comment]
The intent of the code does seem to be that the difference from `zeroTime` is what is supposed to be a multiple of `slideDuration`. From my understanding I agree with your change. Let's see what the tests say.

**Comment by srowen** [comment]
ok to test

**Comment by viirya** [comment]
You will not directly call `slice()`. When `slice()` is called internally, e.g. in `WindowedDStream.compute()`, the `Interval` given as parameter to `slice` is valid.

**Comment by wesleymiao** [comment]
@viirya , you are right. I've changed a little bit to only perform floor when the fromTime and toTime are not aligned relative to zeroTime.

Also fixed a typo in the log message.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 1]
```diff
@@ -762,16 +762,22 @@ abstract class DStream[T: ClassTag] (
     if (!isInitialized) {
       throw new SparkException(this + " has not been initialized")
     }
-    if (!(fromTime - zeroTime).isMultipleOf(slideDuration)) {
-      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
-        + slideDuration + ")")
+
+    val alignedToTime = (toTime - zeroTime).isMultipleOf(slideDuration) match {
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: I think match syntax for booleans is needlessly complicated. You can just write a more straightforward `if`.

Comment (wesleymiao) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: changed to if-else.

File: streaming/src/main/scala/org/apache/spark/streaming/Time.scala [hunk 2]
```diff
@@ -63,6 +63,11 @@ case class Time(private val millis: Long) {
     new Time((this.millis / t) * t)
   }
 
+  def floor(that: Duration, zeroTime: Time): Time = {
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/Time.scala]: Could you also add some tests for this new method?

**Comment by srowen** [comment]
A little test would be great, yeah. Would still love a quick look from @tdas to make sure we're not missing something.

File: streaming/src/main/scala/org/apache/spark/streaming/Time.scala (hunk 2 above)
Comment (wesleymiao) [inline streaming/src/main/scala/org/apache/spark/streaming/Time.scala]: Added a few tests.

**Comment by shaneknapp** [comment]
i'll retrigger this job once i restart jenkins

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by wesleymiao** [comment]
Hi, @tdas , will you be able to take a look at this PR? My team in Autodesk is trying to adopt Spark Streaming to do some real-time analytics for our services' API logs to serve various purposes, among which is to monitor our services' health in real-time based on API log statistics. Then we run into this situation where we need to perform reduceByKeyAndWindow on an already windowed DStream. 

What I'd like to achieve is to do multiple-level reduce on the source of API logging stream. For example - let's say the source logging stream is at interval 1 second. The first level aggregation would be every 1 minute, which is the 60 intervals of the source stream. The second level would be every 1 hour. The third level would be 1 day. And we can do more levels if we want.

What I hope is that at each level we'll do reduceByKeyAndWindow so that its aggregation can be done over its immediate previous level, instead of always aggregating over the source stream. Level 3's reduceByKey will be based on level 2's result, level 2 is based one level 1 and level 1 is based on the source stream.

I would think this approach will be more efficient than always reducing over the source stream, particularly for the higher level (like daily and weekly) aggregation.

**Comment by srowen** [comment]
I'm proceeding with this since everywhere in the code it appears that the intent is for RDDs to start an end on a multiple of the slide duration _after zero time_. Therefore it seems like the intent must have been to take into account `zeroTime`.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala [hunk 3]
```diff
@@ -762,16 +762,22 @@ abstract class DStream[T: ClassTag] (
...
     }
-    if (!(toTime - zeroTime).isMultipleOf(slideDuration)) {
-      logWarning("toTime (" + fromTime + ") is not a multiple of slideDuration ("
-        + slideDuration + ")")
+
+    val alignedFromTime = if ((fromTime - zeroTime).isMultipleOf(slideDuration)) {
+      fromTime
+    } else {
+      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
+      + slideDuration + ")")
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala]: Incorrect indentation.

File: streaming/src/main/scala/org/apache/spark/streaming/Time.scala (hunk 2 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/Time.scala]: Should this be made public? If so, would be good to add some docs.

**Comment by tdas** [comment]
Aah, this has already been merged before I got to it. That's okay, barring a few nits, it was good. Thanks @wesleymiao @srowen

**Pull request closed without merging**